java-checks/src/main/java/org/sonar/java/checks/InterruptedExceptionCheck.java

```diff
@@ -148,9 +149,45 @@ private static boolean isFinallyInterrupting(@Nullable BlockTree blockTree) {
     return blockVisitor.threadInterrupted;
   }
 
-  private static boolean throwInterruptedException(Symbol.MethodSymbol symbol) {
-    return !symbol.isUnknown()
-      && symbol.thrownTypes().stream().anyMatch(INTERRUPTING_TYPE_PREDICATE);
+  private static boolean isInterruptingExceptionExpression(ExpressionTree expressionTree) {
+    return INTERRUPTING_TYPE_PREDICATE.test(expressionTree.symbolType());
+  }
+
+  private static class InterruptingStatementCollector extends MethodTreeUtils.MethodInvocationCollector {
+
+    public InterruptingStatementCollector() {
+      super(
+        symbol -> symbol.thrownTypes().stream().anyMatch(INTERRUPTING_TYPE_PREDICATE)
+      );
+    }
+
+    @Override
+    public void visitTryStatement(TryStatementTree tryStatementTree) {
+      // If inner `try` statement catches interrupting types: cut analysis of its try block, because possible
+      // interruptions thrown there are handled within the scope of the catch block.
+      // Yet, all other blocks (resources, catch, finally) of inner `try`s must still be analyzed, because possible
+      // interruptions thrown there must be handled within the scope of the outer try.
+
+      boolean isCatchingAnyInterruptingTypes = tryStatementTree.catches().stream().anyMatch(catchTree ->
+        getCaughtTypes(catchTree.parameter()).stream().anyMatch(INTERRUPTING_TYPE_PREDICATE)
+      );
+
+      scan(tryStatementTree.resourceList());
+      if (!isCatchingAnyInterruptingTypes) {
+        scan(tryStatementTree.block());
+      }
+      scan(tryStatementTree.catches());
+      scan(tryStatementTree.finallyBlock());
+    }
+
+    @Override
+    public void visitThrowStatement(ThrowStatementTree tree) {
+      // besides to method invocation, we also need to collect throw statements
```

Comment:
While you're right that we should fix it, I'd prefer if you split it into 2 separate commits and have a new ticket. You can do it in 1 PR, just 2 separate commits so we keep track of the issue correctly